src/nmea.cpp

```diff
@@ -0,0 +1,1043 @@
+{
+	/* timeout additional to poll */
+	uint64_t time_started = gps_absolute_time();
+
+	while (true) {
+		/* pass received bytes to the packet decoder */
+		while (_bytes_parsed < _bytes_readed) {
+			int l = 0;
+
+			if ((l = parseChar(_read_buf[_bytes_parsed])) > 0) {
+				/* return to configure during configuration or to the gps driver during normal work
+				 * if a packet has arrived */
+				int ret = handleMessage(l);
+
+				if (ret > 0) {
+					return ret;
+				}
+			}
+
+			_bytes_parsed++;
+		}
+
+		/* everything is read */
+		_bytes_parsed = _bytes_readed = 0;
+
+		/* then poll or read for new data */
+		int ret = read(_read_buf, sizeof(_read_buf), timeout);
+
+		if (ret < 0) {
+			/* something went wrong when polling */
+			return -1;
+
+		} else if (ret == 0) {
+			/* Timeout while polling or just nothing read if reading, let's
+			 * stay here, and use timeout below. */
+
+		} else if (ret > 0) {
+			/* if we have new data from GPS, go handle it */
+			_bytes_readed = ret;
+		}
+
+		/* in case we get crap from GPS or time out */
+		if (time_started + timeout * 1000 * 2 < gps_absolute_time()) {
+			return -1;
+		}
+	}
+}
+
+#define HEXDIGIT_CHAR(d) ((char)((d) + (((d) < 0xA) ? '0' : 'A'-0xA)))
+
+int GPSDriverNMEA::parseChar(uint8_t b)
+{
+	int iRet = 0;
+
+	switch (_decode_state) {
+	/* First, look for sync1 */
+	case NMEADecodeState::uninit:
+		if (b == '$') {
+			_decode_state = NMEADecodeState::got_sync1;
+			_rx_buffer_bytes = 0;
+			_rx_buffer[_rx_buffer_bytes++] = b;
+		}
+
+		break;
+
+	case NMEADecodeState::got_sync1:
+		if (b == '$') {
+			_decode_state = NMEADecodeState::got_sync1;
+			_rx_buffer_bytes = 0;
+
+		} else if (b == '*') {
+			_decode_state = NMEADecodeState::got_asteriks;
+		}
+
+		if (_rx_buffer_bytes >= (sizeof(_rx_buffer) - 5)) {
+			_decode_state = NMEADecodeState::uninit;
+			_rx_buffer_bytes = 0;
+
+		} else {
+			_rx_buffer[_rx_buffer_bytes++] = b;
+		}
+
+		break;
+
+	case NMEADecodeState::got_asteriks:
+		_rx_buffer[_rx_buffer_bytes++] = b;
+		_decode_state = NMEADecodeState::got_first_cs_byte;
+		break;
+
+	case NMEADecodeState::got_first_cs_byte:
+		_rx_buffer[_rx_buffer_bytes++] = b;
+		uint8_t checksum = 0;
+		uint8_t *buffer = _rx_buffer + 1;
+		uint8_t *bufend = _rx_buffer + _rx_buffer_bytes - 3;
+
+		for (; buffer < bufend; buffer++) { checksum ^= *buffer; }
+
+		if ((HEXDIGIT_CHAR(checksum >> 4) == *(_rx_buffer + _rx_buffer_bytes - 2)) &&
+		    (HEXDIGIT_CHAR(checksum & 0x0F) == *(_rx_buffer + _rx_buffer_bytes - 1))) {
+			iRet = _rx_buffer_bytes;
+		}
+
+		decodeInit();
+		break;
+	}
+
+	return iRet;
+}
+
+void GPSDriverNMEA::decodeInit()
+{
+	_rx_buffer_bytes = 0;
+	_decode_state = NMEADecodeState::uninit;
+}
+
+int GPSDriverNMEA::configure(unsigned &baudrate, const GPSConfig &config)
+{
+	if (config.output_mode != OutputMode::GPS) {
+		GPS_WARN("NMEA: Unsupported Output Mode %i", (int)config.output_mode);
+		return -1;
+	}
+	setBaudrate(baudrate);
```

Comment:
`baudrate` might be 0, in which case there should be auto-detection. If not supported, you can just set `baudrate` to `NMEA_DEFAULT_BAUDRATE` if it's 0.

Reply:
I had some problems with the original code and removed the auto-detection. I will try again and test the code or add your suggestions.

Reply:
Now the automatic detection of baudrate works if `baudrate` is 0. Also, the fallback is the `NMEA_DEFAULT_BAUDRATE` if no GPS is detected (except if specified).